docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml

```diff
@@ -11,6 +11,10 @@ spec:
   endpoints:
     - path: /metrics
       port: https
+      scheme: https
+      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+      tlsConfig:
+        insecureSkipVerify: true
   selector:
```

Comment:
I do not think that we should move forward here. See the doc: https://book.kubebuilder.io/reference/metrics.html

These metrics are protected by kube-auth-proxy by default if using kubebuilder. Kubebuilder v2.2.0+ scaffold a clusterrole which can be found at config/rbac/auth_proxy_client_clusterrole.yaml.

```
You will need to grant permissions to your Prometheus server so that it can scrape the protected metrics. To achieve that, you can create a clusterRoleBinding to bind the clusterRole to the service account that your Prometheus server uses.

You can run the following kubectl command to create it. If using kubebuilder <project-prefix> is the namePrefix field in config/default/kustomization.yaml.

kubectl create clusterrolebinding metrics --clusterrole=<project-prefix>-metrics-reader --s
```

Did you follow that?

Reply:
/hold

Reply:
Yes, but I'm not sure how your comment is related.

The proxy is [by default](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/default/manager_auth_proxy_patch.yaml#L15) exposed only via `https` so scraping using `scheme: http` [will result in a 400 error](https://github.com/golang/go/blob/520f3b72db7befab2028d9a47376267cf2d274a9/src/net/http/server.go#L1844-L1845). Annoyingly this directly implemented in golang, so the rbac proxy doesn't log anything and prometheus doesn't log the response body.

The `bearerTokenFile` is required for prometheus to actually [include a](https://github.com/prometheus-operator/prometheus-operator/blob/da324dd3d75ce7db24c6c55704a2231d1249fec6/pkg/prometheus/promcfg.go#L916) `Authorization` [header](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).

This header is required so the proxy can actually validate whether the request has the necessary permissions (you get a [401 otherwise, while the missing binding would result in a 403](https://github.com/brancz/kube-rbac-proxy/blob/e4b31758aedb3d29d8f50d615828401a30d28c1e/pkg/proxy/proxy.go#L62))

The `insecureSkipVerify` is required as the rbac proxy uses a self-generate certificate in the current configuration https://github.com/brancz/kube-rbac-proxy/blob/e4b31758aedb3d29d8f50d615828401a30d28c1e/main.go#L261

@estroz proposed to use the webhook certificate https://github.com/kubernetes-sigs/kubebuilder/pull/2065#discussion_r588711493, but this seems out of scope of this PR.

Reply:
I can be missing something here our misunderstanding. However, see: https://github.com/kubernetes-sigs/kubebuilder/pull/1317/files#diff-20a08dcfd7ad4ed3343d00cf32c7844ce41a2d8d5f5df5b6de56852a510c5517R46 from the PR: https://github.com/kubernetes-sigs/kubebuilder/pull/1317 which via the rbac rules allows accessing metrics behind kube-rbac-proxy. 

Then, see that in the e2e tests that are checked: 

- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L165-L170
- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L272-L303

So, I am trying here understand why in the e2e test we get HTTP 200 from the metrics endpoint and you are facing 400 or 403. What is missing? What has been working our e2e tests and not for you? 

Is it because we are getting the [token](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L265-L270) to check and doing the handshake and you are not?  

Then, I understand that the default configuration has been allowing Prometheus to get the metrics. See that the problem point out by you was solved in the past with the rbac roles https://github.com/kubernetes-sigs/kubebuilder/pull/1317.

Anyway, shows that also has some difference between using Prometheus Operator instead of kube-prometheus which has been solved with the solution applied here as you point out : https://github.com/kubernetes-sigs/kubebuilder/issues/1253#issuecomment-597640633 

So, I will 

/hold cancel 

However, would be very nice to have a new issue with the steps required to reproduce the scenario for the future we are able to check that. Describing the cluster and Prometheus used, and if the rbac roles for allowing the Prometheus get the metrics were or not used, and what are the steps to check the error.

Reply:
Actually, Camila's point is pretty sound. If e2e tests are working and this is being tested, we need an issue describing the steps to get this error so that we can reproduce and check that it is an actual bug and not just an error on your side.

Reply:
Looking at the e2e tests, it seems they only test the service endpoint using curl, but the service monitor itself is not covered. The service monitor instructs how prometheus is supposed to scrape the service endpoint.

The test itself is using https (which is configured via scheme), `-k` which skips the certificate validation and injects a token into the request, which seems to be confirming the service monitor misconfiguration.

I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

Reply:
> The test itself is using https (which is configured via scheme), `-k` which skips the certificate validation and injects a token into the request, which seems to be confirming the service monitor misconfiguration.

That actually makes sense. I verified it and we are exactly testing it as you describe.

> I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

The reproduction we were talking about, at least from my part, was how to reproduce the error you fix here. But changing the `curl` approach for a real e2e test would be pretty good to avoid making the same mistake in the future. Why do you think that requiring prometheus is not suitable for e2e?

Reply:
> I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

We install Prometheus in the e2e tests see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L45-L46 

So, I think we need also cover this scenario to be sure that indeed this solution will fix it and we will not break that in the future. Could you help with that?

Reply:
I'm happy to see what can be done, currently I'm struggling with getting the e2e test suite running on the new M1 MacBook, so an update here might require a bit more time.

Reply:
Go 1.16? Not supported yet. There are some modules related changes (like some commands not tidying up go.mod anymore) that haven't been taking into account in kubebuilder nor any of their dependencies.

Reply:
Ah that is unfortunate. If I have some time I can update e2e code after this is merged.

Reply:
Just to highlight: We might be unable to reproduce the issue in the e2e as it now if it is specific only when to use Prometheus Operator instead of kube-prometheus.

Reply:
If you create an issue describing the need of an e2e test that covers this case, I'm ok merging this.

Reply:
Opened https://github.com/kubernetes-sigs/kubebuilder/issues/2087